File: dcooling.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <search.h>

int graph_len;
int max_pos;
int bad_nodes   = 0;
int stop_vertex = -1;
int numpaths    = 0;

void checkPath(int **graph, int path[], int pos) {
 
  // Check if we came to a desired position
  if (path[pos-1] == stop_vertex) {
    if (pos == max_pos) {
      numpaths++;
      printf("Paths found so far: %d\n", numpaths);
    }
    return;
  }

  int v;
  for (v = 0; v < graph_len; v++) {
  
    // If this element is adjancent to the previous one
    // ...go this path recursively
    if (graph[ path[pos-1] ][v] == 1) {

      // Check if vertex is already added
      int i;
      int cont = 0;
      for (i = 0; i < pos; i++) {
        if (path[i] == v) {
          cont = 1;
          break;
        }
      }
      if (cont == 1) continue;

      path[pos] = v;
      checkPath(graph, path, pos+1);
      path[pos] = -1;
    }
  }

}

int main(void) {

  int width, height;

  // Ask for field's width and height
  printf("Width and height? ");
  scanf("%1i %1i", &width, &height);

  if ((width < 1 || width > 10) ||
      (height < 1 || height > 10)) {
    printf("ERROR: Width/height are out of range\n");
    return -1;
  }

  // Create playing field
  char field[width][height];
  int i, j;

  char *line;

  line = (char *)malloc(width*2 * sizeof *line);

  for (i = 0; i < height; i++) {
    printf("Line %2d? ", i);

    scanf(" %20[^\n]s", line);
    
    if ((int)strlen(line) != width*2 - 1) {
      printf("ERROR: You should enter exact number of values");
      free(line);
      return -1;
    }

    for (j = 0; j < width*2; j++) {
      field[i][j] = line[j*2] - '0';
    }
  }

  free(line);

  // Initialize graph
  graph_len = width*height;
  int **graph;

  // Allocate an array
  graph = (int **)malloc(graph_len * sizeof *graph);
  for (i = 0; i < graph_len; i++) {
    graph[i] = (int *)malloc(graph_len * sizeof *graph[i]);
  }

  // Fill-in default values
  for (i = 0; i < graph_len; i++) {
    for (j = 0; j < graph_len; j++) {
      graph[i][j] = 0;
    }
  }

  // Build graph
  int start_vertex   = -1;
  int current_vertex = -1;

  for (i = 0; i < height; i++) {
    for (j = 0; j < width; j++) {
 
      if (field[i][j] != 1) {
 
        // Calculate vertex number
        current_vertex = i*width + j;

        // Find start and stop vertices
        if (field[i][j] == 2) start_vertex = current_vertex;
        if (field[i][j] == 3) stop_vertex = current_vertex;

        //graph[current_vertex][current_vertex] = 1;

        if (i+1 < height && (field[i+1][j] == 0 || field[i+1][j] == 3)) {
          graph[current_vertex][(i+1)*width+j] = 1;
        }

        if (i-1 >= 0 && (field[i-1][j] == 0 || field[i-1][j] == 3)) {
          graph[current_vertex][(i-1)*width+j] = 1;
        }

        if (j+1 < width && (field[i][j+1] == 0 || field[i][j+1] == 3)) {
          graph[current_vertex][i*width+(j+1)] = 1;
        }

        if (j-1 >= 0 && (field[i][j-1] == 0 || field[i][j-1] == 3)) {
          graph[current_vertex][i*width+(j-1)] = 1;
        }
      
      }
      else {
        bad_nodes++;
      }
    }
  }


  // DEBUG: print graph matrix and additional info
  /*
  printf("    ");
  for (j = 0; j < graph_len; j++) {
    printf("%0.2d ", j);
  }
  printf("\n");
  for (i = 0; i < graph_len; i++) {
    printf("%0.2d: ", i);
    for (j = 0; j < graph_len; j++) {
      printf("%2d ", graph[i][j]);
    }
    printf("\n");
  }

  printf("Bad Nodes: %d\n", bad_nodes);
  printf("Start Vertex: %d\n", start_vertex);
  printf("Stop Vertex: %d\n", stop_vertex);
  */  

  // Initialize path
  int path[graph_len];

  for (i = 0; i < graph_len; i++) {
    path[i] = -1;
  }
  
  // Add initial vertex
  path[0] = start_vertex;

  // Max position
  max_pos = graph_len - bad_nodes;

  // Backtracing
  checkPath(graph, path, 1);

  printf("Paths: %d\n", numpaths);

  // Deallocate an array
  for (i = 0; i < graph_len; i++) {
    free(graph[i]);
  }
  free(graph);

  return 0;

}

File: dcooling.py
#!/usr/bin/python -tt

from copy import deepcopy
import numpy as np

class CoolingPipe:
  
  def  __init__(self):
    self.path = []
    self.numpaths = 0
    self.start_vertex = -1
    self.end_vertex = -1

    # Temporary
    self.paths = []

    # Get self.width and self.height
    wh = raw_input('Width and height? ')
    self.width, self.height = map(int, wh.split())
    if self.width not in range(2, 11) or self.height not in range(2, 11):
      raise ValueError('Width/height are out of range')

    # Get path scheme
    field_txt = []
    for line in range(1, self.height + 1):

      curline = raw_input('Line {0:2d}? '.format(line))
      values = map(int, curline.split())
      
      if len(values) != self.width:
        raise ValueError('You must enter exactly %s values' % self.width)

      for value in values:
        if value not in range(0, 4):
          raise ValueError('Scheme must consist of 0s to 3s only')

      field_txt.append(curline)

    # Check for correctness
    #print map(lambda x: map(lambda y: y.split().index(2), x.split()), self.prepath) 
    #print map(lambda x: map(lambda y: '0' in y.split(), x.split()), self.prepath) 
    #print any('0' in x.split() for x in self.prepath)
    
    if sum(map(lambda x: x.split().count('2'), field_txt)) != 1:
      raise ValueError('You must have only one intake')

    if sum(map(lambda x: x.split().count('3'), field_txt)) != 1:
      raise ValueError('You must have only one air conditioner')

    # Compose a full path as a multidimensional list
    self.field = map(lambda x: map(int, x.split()), field_txt)
    

  def calculate_path(self):

    # Build Graph
    f = self.field

    nodes = self.height*self.width
    g = np.zeros( (nodes, nodes), dtype=np.int16 )

    for x in range(self.height):
      for y in range(self.width):

        if f[x][y] != 1:

          current_vertex = x*self.width + y

          # Get start vertex
          if f[x][y] == 2:
            self.start_vertex = current_vertex

          # Get final vertex
          if f[x][y] == 3:
            self.end_vertex = current_vertex
          
          if (x+1 in range(self.height) and f[x+1][y] != 1):
            g[current_vertex][(x+1)*self.width+y] = 1
          if (x-1 in range(self.height) and f[x-1][y] != 1):
            g[current_vertex][(x-1)*self.width+y] = 1
          if (y+1 in range(self.width) and f[x][y+1] != 1):
            g[current_vertex][x*self.width + (y+1)] = 1
          if (y-1 in range(self.width) and f[x][y-1] != 1):
            g[current_vertex][x*self.width + (y-1)] = 1

    # Remove disconnected vertices
    # ...remove zero rows
    g = g[~np.all(g == 0, axis = 1)]
    # ...remove zero columns
    gmask = np.all(g == 0, axis = 0)
    self.graph = g[:, ~gmask]

    #print "Start: " + str(self.start_vertex)
    #print "End: " + str(self.end_vertex)
    print self.graph

    # Backtracing
    self.graph_len = len(self.graph)
    self.path.append(self.start_vertex)
    self.__check_path()

    #self.__draw_path()

    return self.path

  #@profile
  def __check_path(self):

    # Temporary
    #if self.numpaths > 11:
    #  return False

    # Check if we came to the desired position
    if self.path[-1] == self.end_vertex:
      # The graph is complete
      if len(self.path) == self.graph_len:
        return True
      # Optimization check for better speed
      # If we're at the final vertex, but unused vertices remain
      else:
        return False

    for v in range(self.graph_len):

      # Check if vertex is already added
      if v in self.path:
        continue

      # Find all neighbors
      if self.graph[ self.path[-1] ][v] == 1:
 
        self.path.append(v)
        if self.__check_path() == True:
          #self.paths.append(deepcopy(self.path))
          self.numpaths += 1
          print "Paths found so far: " + str(self.numpaths)

        # If it leads nowhere - remove it from stack
        self.path.pop()

    return False
      
#  def __draw_path(self):

#    for path in self.paths:
#      c = deepcopy(self.field)

#      for node in path:
#        x, y = node.coords
#        c[x][y] = 8
#        for line in c:
#          print line
#        print "==="

#      print "+++"


cp = CoolingPipe()

print "Paths: ", cp.calculate_path()
#for graph in cp.calculate_path():
#  print "(X, Y): " + str(graph.coords)
#  print "Neighbors: " + str(graph.neighbors)
#  print "==="
